File: python-algorithm/leetcode/problem_775.py
"""775. Global and Local Inversions
https://leetcode.cn/problems/global-and-local-inversions/
"""
from typing import List


class Solution:
    def isIdealPermutation(self, nums: List[int]) -> bool:
        n = len(nums)
        if n <= 2:
            return True
        max_idx = 0
        for i in range(2, n):
            if nums[i] < nums[max_idx]:
                return False
            if nums[i - 1] > nums[max_idx]:
                max_idx = i - 1
        return True

File: java-algo/src/main/java/com/leetcode/Problem775.java
package com.leetcode;

/**
 * 775. Global and Local Inversions
 * https://leetcode.com/problems/global-and-local-inversions/
 */
public class Problem775 {
    public boolean isIdealPermutation(int[] nums) {
        int n = nums.length;
        if (n < 3) {
            return true;
        }
        int maxIdx = 0;
        for (int i = 2; i < n; i++) {
            if (nums[i] < nums[maxIdx]) {
                return false;
            }
            if (nums[i - 1] > nums[maxIdx]) {
                maxIdx = i - 1;
            }
        }
        return true;
    }
}
